File: go/mariani-silver/imageDimensions.go
package main

type ImageDimensions struct {
	width       uint
	height      uint
	topLeft     complex64
	bottomRight complex64
	scaleX      float32
	scaleY      float32
}

func NewImageDimensions(width uint, height uint, topLeft complex64, bottomRight complex64) ImageDimensions {
	scaleX := (real(bottomRight) - real(topLeft)) / float32(width)
	scaleY := (imag(bottomRight) - imag(topLeft)) / float32(height)
	return ImageDimensions{
		width, height,
		topLeft, bottomRight,
		scaleX, scaleY,
	}
}

func (imageDimensions ImageDimensions) SampleTo(x, y uint) complex64 {
	scaledX := float32(x)*imageDimensions.scaleX + real(imageDimensions.topLeft)
	scaledY := float32(y)*imageDimensions.scaleY - imag(imageDimensions.bottomRight)
	return complex(scaledX, scaledY)
}

func (imageDimensions ImageDimensions) SampleToIndex(index uint) complex64 {
	return imageDimensions.SampleTo(
		index%imageDimensions.width,
		index/imageDimensions.width,
	)
}

File: rust/mariani-silver/src/image_dimensions.rs
pub mod image_dimensions {
	use crate::complex::complex::Complex;

	pub struct ImageDimensions {
		pub width: usize,
		pub height: usize,
		pub top_left: Complex,
		pub bottom_right: Complex,
		scale_x: f64,
		scale_y: f64,
	}

	impl ImageDimensions {
		pub fn new(
			width: usize, height: usize,
			top_left: Complex, bottom_right: Complex,
		) -> ImageDimensions {
			let scale_x = (bottom_right.real - top_left.real) / width as f64;
			let scale_y = (bottom_right.imaginary - top_left.imaginary) / height as f64;
			ImageDimensions {
				width,
				height,
				top_left,
				bottom_right,
				scale_x,
				scale_y,
			}
		}

		pub fn sample_to(&self, x: usize, y: usize) -> Complex {
			let scaled_x = x as f64 * self.scale_x + self.top_left.real;
			let scaled_y = y as f64 * self.scale_y - self.bottom_right.imaginary;
			Complex::new(scaled_x, scaled_y)
		}
	}
}